assumeutxo: keep cache when flushing snapshot (#17487 followup)

jamesob opened the pull request:
Make use of #17487.

Benchmark results available here:
  https://github.com/bitcoin/bitcoin/pull/17487#issuecomment-557226923

jamesob committed:
validation: optimization: use Sync() when flushing snapshot to disk

Benchmark results available here:
  https://github.com/bitcoin/bitcoin/pull/17487#issuecomment-557226923

Sjors commented on src/validation.cpp:
```diff
@@ -5033,15 +5033,19 @@ bool ChainstateManager::ActivateSnapshot(
 {
     LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE(
         strprintf("%s (%.2f MB)",
                   snapshot_loaded ? "saving snapshot chainstate" : "flushing coins cache",
```
nit: `(erase ? "syncing" : "flushing") + "coins cache"`

Sjors reviewed (approved):
utACK d6f54dac8dee776188fd88d725e2728e0822b33f

Based on testing #17487 I think it's safe. I'd like to take this for a spin on a rebase of #15606 (the bits that enable `loadtxoutset`), to see if I can reproduce the performance improvement. But imo that doesn't have to hold back this PR.

jamesob committed:
validation: AbortNode if coins_cache disk sync fails

We've historically done this check in most places, so make sure we're
doing it uniformly.

jamesob commented:
Thanks for the look @Sjors. I also realized that there were two outlier calls where we weren't asserting that Flush()/Sync() has succeeded, so I've fixed those.

jamesob commented on src/validation.cpp:
```diff
@@ -5036,12 +5036,12 @@ bool ChainstateManager::ActivateSnapshot(
     LOG_TIME_MILLIS_WITH_CATEGORY_MSG_ONCE(
-        strprintf("%s (%.2f MB)",
-                  snapshot_loaded ? "saving snapshot chainstate" : "flushing coins cache",
+        strprintf("%s coins cache (%.2f MB)",
```
Sounds good, fixed.

jamesob reviewed (commented):
Thanks for the feedback @Sjors.

Sjors commented:
re-utACK 937016b1917a0a07fa579c96fc797237f5027cd1

Sjors commented on src/validation.cpp:
```diff
@@ -5033,15 +5035,19 @@ bool ChainstateManager::ActivateSnapshot(
+    if (snapshot_loaded ? coins_cache.Sync() : coins_cache.Flush()) {
+        return true;
+    } else {
+        return AbortNode("Failed to write to coin database");
```
Same question about `AbortNode` vs. `error` though.

jamesob replied:
`AbortNode()` is used elsewhere in the function that this function is called from, although I did change another reference that you pointed out what inconsistent.

Sjors commented:
From my testing of #15606 - without this PR - I get the impression that flushing isn't working, specifically that RAM is not freed up. It could just be confusion on my end from rebase hell, but I'd like to make sure we're not obscuring a bug by merging this.

jamesob commented:
@Sjors how did you detect that RAM *wasn't* being freed? Did you verify that the flush did not happen based on an absence of logs? 

Measures of memory can be deceptive; I remember bitcoinperf spuriously detecting an increase in memory usage because once claimed, the process' resident set size (RSS) doesn't necessarily go down upon free.

jamesob commented:
Actually this _is_ broken at the moment, since although we don't flush after snapshot load, we do flush (and not sync) in the MaybeRebalanceCaches() -> FlushStateToDisk() call right after activating the chainstate snapshot. Will close this and think about how to rework it.
